### ISSUE TITLE
Configurations GUI

### ISSUE THREAD

**Issue by franzlst** [DLR-RM/RAFCON#53]
Till now the configurations (core, gui, network) has to be modified in ~/.config/rafcon. A GUI would speedup the configuration modification.

Originally created by @sebastian-brunner  (Sebastian.Brunner@dlr.de) at 2016-02-22 13:37:57+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
As long as the configuration of RAFCON grows more and more complex a visual interface to configure RAFCON will gain overview and control, especially for inexperienced users.

The widget may also offers options to export the configuration to a specific path-location or to load new configuration files from other path-locations.

For the LIBRARY_PATHS the widget would offer options to re-configure mounted library sets.


Originally created by @Rbelder  at 2016-03-01 20:19:47+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Added configuration GUI by commit b14ee4b. (menu_bar -> properties)
- when applying, all settings will be stored into the specific config file
- various changes require a restart
- libraries and shortkeys will be updated

missing:
- export and import of new config file


Originally created by Benno  at 2016-08-19 09:06:47+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
I have done some refactoring and sort the entries to help the human find its targets. :smile: 

Some refactoring still can be done to maybe specify exactly when a restart is needed. 

This also can be reduced when the issue #160 is implemented and its feature is used by all core and gui elements to observe changes on there parametrization.

Anyhow the import and export of config files and a more specific dialog window can finish the issue for now.


Originally created by @Rbelder  at 2016-08-23 10:22:15+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Sebastian and I determined that the LIBRARIES, the SHORTCUTS and the CHECK_PYTHON_FILES_WITH_PYLINT are the only configs he is sure that need no restart. But what i actually do is to set each config with set_config() without checking if a restart is needed or not. Maybe this can cause big troubles...
btw. did you notice that if you change the gaphas setting and open a new SM without restart, you can run both styles in one RAFCON ;)  


Originally created by Benno  at 2016-08-24 07:55:53+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
All stuff dependent on a state machine and its model like enabling auto backup and history  (or parameters of those) will be used in reopened or newly created state machines. So there is barely no setting parameter which is not updateable on the fly as long as all state machines are saved.

An example of  parameters not even necessary to update are the LOGGING_SHOW\* parameters.

And by now I don't find any parameter which is not updated if a refresh all is performed. The User only has to be given the choice between a refresh all, refresh library and so on dependent on the changed parameter. 

Actually the both style stuff is not a feature. ;)

So we can site down together defining for every parameter which action is needed to apply it. That will create a tuple of length three for every parameter. But for the details of my thoughts we should have 10-15min discussion.


Originally created by @Rbelder  at 2016-08-24 08:41:03+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Refactored in commit 2fcb95c.
- still needed to implement an action which should be triggered when changing Logger settings (Rico suggested to set the logger checkboxes)
- not implemented a redrawing of GAPHAS when changing SHOW_NAMES_ON_DATA_FLOWS or ROTATE_NAMES_ON_CONNECTIONS yet
- not implemented restart of open state-editors when changing SOURCE_EDITOR_STYLE yet
- action triggered by following changes has to be advised by Franz:
  - MAX_VISIBLE_LIBRARY_HIERARCHY
  - MINIMUM_SIZE_FOR_CONTENT
  - ENABLE_CACHING
- Import/ export feature is still missing


Originally created by Benno  at 2016-08-26 08:57:24+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Refactored in commit d3aed21. 
- changing the LOGGING_SHOW_\* settings will check/ uncheck the checkboxes 
- changing the SOURCE_EDITOR_STYLE reloads the source_editor at the root_state ( it would be nicer if all currently open tabs (active and pinned) will be reloaded with the new style)
- importing the main_window_controller instead of giving as parameter

still missing:
- [x] Import/ export feature

actions for: 
- [x] MAX_VISIBLE_LIBRARY_HIERARCHY
- [x] MINIMUM_SIZE_FOR_CONTENT
- [x] ENABLE_CACHING
- [x] SHOW_NAMES_ON_DATA_FLOWS ( redrawing of GAPHAS) 
- [x] ROTATE_NAMES_ON_CONNECTIONS ( redrawing of GAPHAS) 


Originally created by Benno  at 2016-08-29 11:48:31+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Update in commit  c61a44f:
Implemented im- /export feature. 
- Imported configs are automatically applied
- only applied settings will be exported


Originally created by Benno  at 2016-08-31 09:42:40+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Commit 746575f3a047 fixes handling of relative library locations. We also need a unit test for this feature.


Originally created by @sebastian-brunner  (Sebastian.Brunner@dlr.de) at 2016-09-02 14:50:51+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Adding or removing a library path in the GUI is currently not working!


Originally created by @sebastian-brunner  (Sebastian.Brunner@dlr.de) at 2016-09-26 08:22:14+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
- [x] An OK-Button that close the window and apply the changes would be intuitive. At the moment I have to cancel (What is not canceling the already applied changes) or press the destroy cross-button what is not intuitive, too.
- [x] Also helpful would be to change sensitivity of the apply button according changes that can be applied or not
- [x] For the adding and removing problem changes made in 5df8189c and 58158de causing the problem -> maybe a unit test can help -> adding logger warning in both methods like "Library tree view has to be focused for adding libraries" would be helpful.


Originally created by @Rbelder  at 2016-09-26 09:07:47+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Commit 7250f4d should fix the add/ remove problem


Originally created by Benno  at 2016-09-26 14:40:48+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
So I can confirm the remove/add of library path is working again.

I also insert into the setting configuration GUI the two sub-task noted above, see d66442d5c. A OK button was added, the apply button changes it sensitivity to indicate changes and the OK button is applying and closing the window. I fixed that the changed_keys dict was not reset in get_settings method of the model.

@Benno I look through the model and so on. Maybe you refactor the naming of methods and documentation. And maybe look also at the controller, too. One example, you use the method set_properties of the model in the controller while canceling changes which calls get_properties and basically is doing a "reset to global config values". That is maybe a load/reset but not a getter method which (the getter) generally has a return value.


Originally created by @Rbelder  at 2016-09-27 10:25:30+00:00 (moved from RMC internal repository)
